### PR TITLE
Fix int64 overflow in im2col/col2im sliding-window count (#189667)

### LINKED ISSUE #189667: [CUDA] illegal memory write on im2col_kernel
### 🐛 Describe the bug

```python
# Bug: im2col_cuda integer overflow on huge kernel/dilation/padding/stride

import torch
print("torch:", torch.__version__)

torch.cuda.init()

x = torch.zeros(32, 64, 3, 3, device='cuda')
torch.nn.functional.unfold(
    x,
    kernel_size=(2**63 - 2, 2**63 - 2),
    dilation=(1000000000000, 1000000000000),
    padding=(1000000000000, 1000000000000),
    stride=(1000000000000, 1000000000000),
)
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ write of size 4 bytes
=========     at void at::native::im2col_kernel<float>(long, const T1 *, long, long, long, long, long, long, long, long, long, long, long, long, T1 *)+0x1300
=========     by thread (240,0,0) in block (2,0,0)
=========     Address 0x7f8f23400000 is out of bounds
=========     and is 1 bytes after the nearest allocation at 0x7f8f22000000 of size 20971520 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

### PR DESCRIPTION
**Pull request by CodersAcademy006**
Fixes #189667

### Root cause
`sliding_window_count()` in `im2col_shape_check.h` computed
`input_size + 2 * pad - (dilation * (kernel - 1) + 1)` directly in `int64_t`.
Adversarial arguments (e.g. `unfold` with `kernel_size=2**63-2` plus large
`dilation`/`padding`/`stride`) overflow `int64_t` and wrap to a spuriously
valid positive count. That count passes the existing `>= 1` shape check and
then drives an out-of-bounds write in `im2col_kernel` on CUDA (crash under
`compute-sanitizer`).

### Fix
Compute the numerator with checked arithmetic via `c10::mul_overflows` /
`c10::add_overflows` -- the same helper already used a few lines above for the
`kernel_width * kernel_height` product -- and `TORCH_CHECK` on overflow, so the
bad shape is rejected on the host before any kernel launch. Because the guard
lives in the shared helper it covers both the im2col (`unfold`) and col2im
(`fold`) paths, on CPU and CUDA. No new abstraction is introduced.

### Test
Added a CPU-runnable regression case to `test_unfold_invalid_arg`:

```
python test/test_nn.py -k test_unfold_invalid_arg
```

Draft: opening for CI to run the full build + CUDA sanitizer path (CUDA does
not build locally on macOS).

### CODE REVIEW

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!

**Comment by CodersAcademy006** [comment]
Fair, and noted for going forward. For this one specifically, #189667 is already labeled `actionable`/`triaged`/`module: crash` (it's a fuzzer-found illegal memory write, not a feature ask), so happy to keep any further design discussion on that thread if there's anything you'd want changed before this lands.

**Comment by CodersAcademy006** [comment]
Re-pointing this at the right issue. @malfet closed #189667 as a duplicate of #188959, consolidating the fold/unfold overflow reports, so this PR now maps to **#188959**, which is labelled `triaged`/`module: int overflow`/`topic: fuzzer`. That also answers @albanD's point above: the issue was already triaged before the PR, and #188972 (the col2im half) is closed, so nothing else is currently fixing it.

The fix is in `sliding_window_count()` in `aten/src/ATen/native/im2col_shape_check.h`, which is shared by im2col and col2im, so it covers both halves of #188959 in one place.
